ush/python/pygfs/task/analysis.py

```diff
@@ -328,3 +356,69 @@ def tgz_diags(statfile: str, diagdir: str) -> None:
             # Add diag files to tarball
             for diagfile in diags:
                 tgz.add(diagfile, arcname=os.path.basename(diagfile))
+
+
+@logit(logger)
+def find_value_in_nested_dict(nested_dict: Dict, key: str) -> Any:
```

Comment:
As written, this is going to have strange (possible inconsistent) behavior for keys that appear multiple times. If the top-most dict contains the target key, that will be the value returned. Otherwise, it will iterate through keys looking for the target key, but dicts should not be considered naturally ordered (even though the current implementation of python preserves ordering), so if the target key is contained in multiple nested dicts (but not the top one), the value retrieved could change from execution to execution.

Even without the undetermined ordering issue, the retrieved value won't necessarily be the shallowest (unless it is in the top-level dict) or the deepest matching key.

Reply:
This is true, and thus, I don't think we can move it to a generic module.
Think of giving someone just a house number to find and leaving them at the edge of town.
Which specific house they find will depend on the choices they make; left, or right.
Same is with this search.  This will return the first key the search encounters.

Perhaps we should remove this method and instead look for `observations` under `cost function`.  If it finds it, then thats it.  If not, it should look for `observations` at the top-level. 
Those are the use cases for now.
If we run into problems, we can revisit this and implement a `find_observations` method targeted for this rather than a generic `find_key_in_nested_dict`.